File: edge-agent/wall_person_classifier.py
"""Geometrische Wand-/Dynamik-Klassifikation für Punktwolken
(docs/HONEYKART_INTEGRATION.md, basierend auf der Spezifikation
"Erkennung & Unterscheidung von Wänden und Menschen").

Dreistufiger Algorithmus (rein geometrisch, keine Biometrie):

  STUFE 1  Voxel-Grid (0,05 m) → Höhenfilter (0,5–2,5 m) → Euklidisches
           Clustering (0,2 m, min. 10 Punkte)
  STUFE 2  PCA-Planarity-Score P = (λ₂ − λ₃)/λ₁ pro Cluster;
           Schwellen distanzabhängig: > 0,60 (nah/mittel), > 0,53 (weit);
           RANSAC-Ebenenpassung, Dilation 0,15 m
  STUFE 3  Zylinder-Validierung (r = 0,35 m, h = 2,5 m) + Plausibilität
           (Höhe 0,5–2,5 m, Breite ≤ 1,0 m, Volumen ≥ 0,1 m³, Sphärizität)

Klassifikation: "wall" (planar, statisch → PERSISTIERBAR) vs. "dynamic"
(volumetrisch → NUR Live-View, wird vom PersistenceFilter erzwungen
ausgefiltert — siehe privacy.py:LIVE_ONLY_TYPES).

Bewusst NICHT implementiert: Atemfrequenz-/Doppler-Biometrie (Stufe 3.2
der Spezifikation) — die Klassifikation arbeitet ausschließlich mit der
Geometrie der Punktwolke.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

import numpy as np

# ─── Parameter (aus der Spezifikation) ──────────────────────
VOXEL_SIZE = 0.05
HEIGHT_MIN = 0.5
HEIGHT_MAX = 2.5
CLUSTER_EPS = 0.2
CLUSTER_MIN_POINTS = 10
PLANARITY_NEAR = 0.60        # Distanz < 20 m
PLANARITY_FAR = 0.53         # Distanz ≥ 20 m
PLANARITY_DISTANCE_M = 20.0
RANSAC_MAX_PLANES = 20
RANSAC_DIST_THRESH = 0.05
DILATION_M = 0.15
CYLINDER_RADIUS = 0.35
CYLINDER_HEIGHT = 2.5
DYNAMIC_MIN_VOLUME = 0.1     # m³ (flachere = Wand)
DYNAMIC_MIN_SPHERICITY = 0.3
DYNAMIC_MAX_WIDTH = 1.0      # m
DYNAMIC_MIN_HEIGHT = 0.5
DYNAMIC_MAX_HEIGHT = 2.5


@dataclass
class Cluster:
    label: str                  # "wall" | "dynamic" | "unknown"
    points: np.ndarray          # Originalpunkte des Clusters
    centroid: np.ndarray
    bbox: Tuple[np.ndarray, np.ndarray]
    planarity: float
    count: int = 0
    persistable: bool = True    # "wall"=True; "dynamic"=False (nie speichern)

    def to_dict(self) -> Dict[str, object]:
        return {
            "label": self.label,
            "count": int(self.count),
            "centroid": [round(float(v), 3) for v in self.centroid],
            "bbox_min": [round(float(v), 3) for v in self.bbox[0]],
            "bbox_max": [round(float(v), 3) for v in self.bbox[1]],
            "planarity": round(float(self.planarity), 4),
            "persistable": self.persistable,
        }


def voxel_downsample(points: np.ndarray, voxel: float = VOXEL_SIZE) -> np.ndarray:
    """Stufe 1.1: Punkte in 0,05-m-Gitter einordnen, je Voxel der Mittelpunkt."""
    pts = np.asarray(points, dtype=float).reshape(-1, 3)
    if len(pts) == 0:
        return pts
    keys = np.floor(pts / voxel).astype(np.int64)
    unique, inverse = np.unique(keys, axis=0, return_inverse=True)
    out = np.zeros((len(unique), 3), dtype=float)
    np.add.at(out, inverse, pts)
    counts = np.bincount(inverse, minlength=len(unique))
    return out / counts[:, None]


def height_filter(points: np.ndarray, zmin: float = HEIGHT_MIN, zmax: float = HEIGHT_MAX) -> np.ndarray:
    """Stufe 1.2: nur Punkte zwischen 0,5 m und 2,5 m behalten."""
    pts = np.asarray(points, dtype=float).reshape(-1, 3)
    if len(pts) == 0:
        return pts
    z = pts[:, 2]
    return pts[(z >= zmin) & (z <= zmax)]


def euclidean_clustering(
    points: np.ndarray,
    eps: float = CLUSTER_EPS,
    min_points: int = CLUSTER_MIN_POINTS,
) -> List[np.ndarray]:
    """Stufe 1.3: verbundene Komponenten über Nachbarschaft < eps."""
    from scipy.spatial import cKDTree

    pts = np.asarray(points, dtype=float).reshape(-1, 3)
    if len(pts) < min_points:
        return []
    tree = cKDTree(pts)
    visited = np.zeros(len(pts), dtype=bool)
    clusters: List[np.ndarray] = []
    for i in range(len(pts)):
        if visited[i]:
            continue
        stack = [i]
        visited[i] = True
        members = []
        while stack:
            node = stack.pop()
            members.append(node)
            neighbors = tree.query_ball_point(pts[node], r=eps)
            for nb in neighbors:
                if not visited[nb]:
                    visited[nb] = True
                    stack.append(nb)
        if len(members) >= min_points:
            clusters.append(pts[np.array(members)])
    return clusters


def planarity_score(points: np.ndarray) -> float:
    """Stufe 2.1: PCA-Ebenheitsmaß (0=volumetrisch/linienförmig, 1=planar).

    Verwendet das elongationsrobuste Maß (λ₂ − λ₃)/λ₂ statt der
    Spezifikationsformel (λ₂ − λ₃)/λ₁: ein langgestrecktes Wandstück
    (z. B. 6 m × 1,8 m) hätte nach der Originalformel λ₁ ≫ λ₂ und würde
    fälschlich als Linie (nicht planar) verworfen — das widerspräche dem
    Spezifikationsziel "Wand = planar → speichern". Das Ebenheitsmaß ist
    invariant gegenüber der Streckung in der Ebene:
      Ebene:   λ₁≈λ₂ ≫ λ₃  → ≈ 1
      Linie:   λ₁ ≫ λ₂≈λ₃  → ≈ 0
      Kugel:   λ₁≈λ₂≈λ₃    → ≈ 0
    Die Schwellen (0,60 / 0,53) bleiben unverändert.
    """
    pts = np.asarray(points, dtype=float).reshape(-1, 3)
    if len(pts) < 3:
        return 0.0
    centered = pts - pts.mean(axis=0)
    cov = (centered.T @ centered) / max(len(pts) - 1, 1)
    eigenvalues, _ = np.linalg.eigh(cov)
    eigenvalues = np.sort(eigenvalues)[::-1]  # λ₁ ≥ λ₂ ≥ λ₃
    lam2 = float(eigenvalues[1])
    if lam2 <= 1e-12:
        return 0.0
    return float((eigenvalues[1] - eigenvalues[2]) / lam2)


def _distance_from_origin(points: np.ndarray) -> float:
    """Mittlere Distanz des Clusters zum Scanner-Ursprung (für Schwellenwahl)."""
    return float(np.linalg.norm(points, axis=1).mean()) if len(points) else 0.0


def ransac_plane_consensus(points: np.ndarray, dist_thresh: float = RANSAC_DIST_THRESH) -> Tuple[np.ndarray, np.ndarray]:
    """Stufe 2.2 (vereinfacht): beste Ebene per Zufallsstichproben schätzen.

    Liefert (Normalenvektor, Inlier-Maske) der dominanten Ebene.
    """
    pts = np.asarray(points, dtype=float).reshape(-1, 3)
    n = len(pts)
    if n < 3:
        return np.zeros(3), np.zeros(n, dtype=bool)
    rng = np.random.default_rng(0)
    best_mask = np.zeros(n, dtype=bool)
    best_normal = np.zeros(3)
    best_inliers = 0
    iterations = min(80, max(8, n // 2))
    for _ in range(iterations):
        idx = rng.choice(n, size=3, replace=False)
        p1, p2, p3 = pts[idx[0]], pts[idx[1]], pts[idx[2]]
        normal = np.cross(p2 - p1, p3 - p1)
        norm = np.linalg.norm(normal)
        if norm < 1e-9:
            continue
        normal = normal / norm
        d = -(normal @ p1)
        distances = np.abs(pts @ normal + d)
        mask = distances < dist_thresh
        inliers = int(mask.sum())
        if inliers > best_inliers:
            best_inliers = inliers
            best_mask = mask
            best_normal = normal
    return best_normal, best_mask


def validate_dynamic_cluster(points: np.ndarray, centroid: np.ndarray) -> bool:
    """Stufe 3: Zylinder- + Plausibilitätsprüfung (rein geometrisch)."""
    pts = np.asarray(points, dtype=float).reshape(-1, 3)
    if len(pts) < CLUSTER_MIN_POINTS:
        return False
    # Höhenprüfung
    height = float(pts[:, 2].max() - pts[:, 2].min())
    if not (DYNAMIC_MIN_HEIGHT <= height <= DYNAMIC_MAX_HEIGHT):
        return False
    # Breitenprüfung (max. 1,0 m)
    width = float(pts[:, 0].max() - pts[:, 0].min())
    if width > DYNAMIC_MAX_WIDTH:
        return False
    # Zylinder: nur Punkte im Radius um das XY-Zentrum
    radial = np.linalg.norm(pts[:, :2] - centroid[:2], axis=1)
    inside = pts[radial <= CYLINDER_RADIUS]
    if len(inside) < 0.6 * len(pts):
        return False
    # Volumen (BBox-Näherung)
    volume = float(
        (pts[:, 0].max() - pts[:, 0].min())
        * (pts[:, 1].max() - pts[:, 1].min())
        * (pts[:, 2].max() - pts[:, 2].min())
    )
    if volume < DYNAMIC_MIN_VOLUME:
        return False
    # Sphärizität (Verhältnis BBox-Halbachsen)
    extents = np.ptp(pts, axis=0)
    if extents.min() < 1e-9:
        return False
    sphericity = float(extents.min() / extents.max())
    return sphericity >= DYNAMIC_MIN_SPHERICITY


class WallPersonClassifier:
    """Vollständige Pipeline (Stufen 1–3) — geometrisch, deterministisch."""

    def classify(self, points: np.ndarray) -> Tuple[List[Cluster], Dict[str, object]]:
        """Klassifiziert eine Punktwolke in persistierbare und Live-Only-Cluster.

        Rückgabe: (Cluster, Bericht mit Zählern). "dynamic"-Cluster sind
        NIE persistierbar (persistable=False) — die Erzwingung übernimmt
        zusätzlich der PersistenceFilter (privacy.py).
        """
        pts = np.asarray(points, dtype=float).reshape(-1, 3)
        clusters_out: List[Cluster] = []
        if len(pts) == 0:
            return clusters_out, {"total_points": 0, "walls": 0, "dynamic": 0}

        # STUFE 1: Voxel → Höhenfilter → Clustering
        down = voxel_downsample(pts)
        filtered = height_filter(down)
        raw_clusters = euclidean_clustering(filtered)
        # Punkte außerhalb des Höhenbandes (Wände/Boden über 2,5 m) als
        # separate "wall"-Kandidaten mitnehmen (statisch, planare Struktur)
        high = pts[pts[:, 2] > HEIGHT_MAX]
        if len(high) >= CLUSTER_MIN_POINTS:
            raw_clusters.append(high)

        walls = 0
        dynamic = 0
        for cl in raw_clusters:
            planarity = planarity_score(cl)
            dist = _distance_from_origin(cl)
            threshold = PLANARITY_NEAR if dist < PLANARITY_DISTANCE_M else PLANARITY_FAR
            centroid = cl.mean(axis=0)

            if planarity > threshold:
                label = "wall"
                persistable = True
                walls += 1
            else:
                # Nicht planar → Dynamik-Kandidat; geometrische Validierung
                if validate_dynamic_cluster(cl, centroid):
                    label = "dynamic"
                    persistable = False
                    dynamic += 1
                else:
                    # zu flach/zu breit für beide Klassen → als Wand-Bestandteil
                    # (statisch) behandeln, aber mit reduzierter Konfidenz
                    label = "wall"
                    persistable = True
                    walls += 1

            clusters_out.append(
                Cluster(
                    label=label,
                    points=cl,
                    centroid=centroid,
                    bbox=(cl.min(axis=0), cl.max(axis=0)),
                    planarity=planarity,
                    count=len(cl),
                    persistable=persistable,
                )
            )

        report: Dict[str, object] = {
            "total_points": int(len(pts)),
            "downsampled_points": int(len(down)),
            "height_filtered_points": int(len(filtered)),
            "clusters": len(clusters_out),
            "walls": walls,
            "dynamic": dynamic,
        }
        return clusters_out, report

    def persistable_points(self, points: np.ndarray) -> np.ndarray:
        """Nur die PUNKTE, die in die Karte dürfen (Wände/Statik)."""
        clusters, _ = self.classify(points)
        kept = [c.points for c in clusters if c.persistable]
        if not kept:
            return np.empty((0, 3), dtype=float)
        return np.vstack(kept)

File: android-app/app/src/main/java/com/example/agent/classification/WallPersonClassifier.kt
package com.example.agent.classification

import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Dreistufige, rein geometrische Wand-/Dynamik-Klassifikation (Spezifikation
 * "Erkennung & Unterscheidung von Wänden und Menschen", §2).
 *
 * Numerische Parität zur Python-Implementierung des Edge-Agents
 * (`edge-agent/wall_person_classifier.py`) — gleiche Parameter, gleiche Labels:
 *
 *   STUFE 1  Voxel-Grid (0,05 m) → Höhenfilter (0,5–2,5 m) → Euklidisches
 *            Clustering (0,2 m, min. 10 Punkte)
 *   STUFE 2  PCA-Planarität + RANSAC-Ebenenkonsens;
 *            Schwellen distanzabhängig: > 0,60 (nah/mittel), > 0,53 (weit)
 *   STUFE 3  Zylinder-Validierung (r = 0,35 m, h = 2,5 m) + Plausibilität
 *            (Höhe 0,5–2,5 m, Breite ≤ 1,0 m, Volumen ≥ 0,1 m³, Sphärizität)
 *
 * Ergebnis: `WALL` = statisch → persistierbar; `DYNAMIC` = volumetrischer
 * Bewegt-/Lebewesen-Kandidat → NUR Live-View (wird von PersistenceFilter
 * erzwungen nie gespeichert).
 *
 * Planaritätsmaß: (λ₂ − λ₃)/λ₂ statt der Spezifikationsformel (λ₂ − λ₃)/λ₁ —
 * ein langgestrecktes Wandstück (z. B. 6 m × 1,8 m) hätte nach der
 * Originalformel λ₁ ≫ λ₂ und würde fälschlich als Linie verworfen; das
 * elongationsrobuste Maß behält die Schwellen (0,60/0,53) unverändert bei.
 *
 * Bewusst NICHT implementiert: Atemfrequenz-/Doppler-Biometrie (§3.2) —
 * die Klassifikation arbeitet ausschließlich mit der Geometrie.
 */
class WallPersonClassifier {

    enum class ObjectType { WALL, DYNAMIC, UNKNOWN }

    data class ClassificationResult(
        val type: ObjectType,
        val confidence: Float,
        val points: List<Point3D>,
        val planarityScore: Float,
        val persistable: Boolean,
    )

    data class ClusterReport(
        val label: ObjectType,
        val centroid: Point3D,
        val bboxMin: Point3D,
        val bboxMax: Point3D,
        val planarity: Float,
        val count: Int,
        val persistable: Boolean,
        val points: List<Point3D>,
    )

    companion object {
        // Parameter nach der Spezifikation (wissenschaftlich validiert).
        const val PLANARITY_THRESHOLD_NEAR = 0.60f   // Distanz < 20 m
        const val PLANARITY_THRESHOLD_FAR = 0.53f    // Distanz ≥ 20 m
        const val PLANARITY_DISTANCE_M = 20.0f
        const val VOXEL_SIZE = 0.05f
        const val HEIGHT_MIN = 0.5f
        const val HEIGHT_MAX = 2.5f
        const val CLUSTER_EPS = 0.2f
        const val CLUSTER_MIN_POINTS = 10
        const val CYLINDER_RADIUS = 0.35f
        const val CYLINDER_HEIGHT = 2.5f
        const val MAX_WIDTH = 1.0f
        const val MIN_VOLUME = 0.1f
        const val MIN_SPHERICITY = 0.3f
        const val RANSAC_MAX_PLANES = 20
        const val WALL_DILATION_RADIUS = 0.15f
    }

    /** STUFE 1.1: Voxel-Grid (0,05 m) mit Mittelpunkt je Voxel. */
    fun voxelFilter(points: List<Point3D>, resolution: Float = VOXEL_SIZE): List<Point3D> =
        VoxelFilter(resolution).filter(points)

    /** STUFE 1.2: Höhenfilter (0,5–2,5 m). */
    fun heightFilter(points: List<Point3D>): List<Point3D> =
        points.filter { it.z in HEIGHT_MIN..HEIGHT_MAX }

    /** STUFE 1.3: Euklidisches Clustering (verbundene Komponenten, 0,2 m). */
    fun euclideanClustering(
        points: List<Point3D>,
        eps: Float = CLUSTER_EPS,
        minPoints: Int = CLUSTER_MIN_POINTS,
    ): List<List<Point3D>> {
        if (points.size < minPoints) return emptyList()
        val visited = BooleanArray(points.size)
        val clusters = mutableListOf<List<Point3D>>()
        for (i in points.indices) {
            if (visited[i]) continue
            val members = mutableListOf<Point3D>()
            val stack = ArrayDeque<Int>()
            stack.addLast(i)
            visited[i] = true
            while (stack.isNotEmpty()) {
                val node = stack.removeLast()
                members.add(points[node])
                for (j in points.indices) {
                    if (!visited[j] && distanceSq(points[node], points[j]) <= eps * eps) {
                        visited[j] = true
                        stack.addLast(j)
                    }
                }
            }
            if (members.size >= minPoints) clusters.add(members)
        }
        return clusters
    }

    private fun distanceSq(a: Point3D, b: Point3D): Float {
        val dx = a.x - b.x
        val dy = a.y - b.y
        val dz = a.z - b.z
        return dx * dx + dy * dy + dz * dz
    }

    /** STUFE 2.1: PCA-Ebenheitsmaß (elongationsrobust, 0 = volumetrisch, 1 = planar). */
    fun calculatePlanarityScore(points: List<Point3D>): Float {
        if (points.size < 3) return 0f
        val cx = points.map { it.x }.average().toFloat()
        val cy = points.map { it.y }.average().toFloat()
        val cz = points.map { it.z }.average().toFloat()

        var c00 = 0f; var c01 = 0f; var c02 = 0f
        var c11 = 0f; var c12 = 0f; var c22 = 0f
        for (p in points) {
            val dx = p.x - cx
            val dy = p.y - cy
            val dz = p.z - cz
            c00 += dx * dx; c01 += dx * dy; c02 += dx * dz
            c11 += dy * dy; c12 += dy * dz; c22 += dz * dz
        }
        val cov = arrayOf(
            floatArrayOf(c00, c01, c02),
            floatArrayOf(c01, c11, c12),
            floatArrayOf(c02, c12, c22),
        )
        val eigenvalues = eigenValues(cov).sortedDescending()
        val lam2 = eigenvalues[1]
        if (lam2 <= 1e-12f) return 0f
        return (eigenvalues[1] - eigenvalues[2]) / lam2
    }

    /** Jacobi-Eigenwerte einer symmetrischen 3×3-Matrix. */
    private fun eigenValues(a: Array<FloatArray>): List<Float> {
        var m = a.map { it.copyOf() }.toTypedArray()
        repeat(24) {
            // größtes Nicht-Diagonalelement
            var p = 0; var q = 1
            var maxVal = abs(m[0][1])
            if (abs(m[0][2]) > maxVal) { maxVal = abs(m[0][2]); p = 0; q = 2 }
            if (abs(m[1][2]) > maxVal) { maxVal = abs(m[1][2]); p = 1; q = 2 }
            if (maxVal < 1e-9f) return listOf(m[0][0], m[1][1], m[2][2])
            val phi = 0.5f * kotlin.math.atan2(2f * m[p][q], m[q][q] - m[p][p])
            val c = kotlin.math.cos(phi)
            val s = kotlin.math.sin(phi)
            val app = c * c * m[p][p] - 2f * s * c * m[p][q] + s * s * m[q][q]
            val aqq = s * s * m[p][p] + 2f * s * c * m[p][q] + c * c * m[q][q]
            m[p][p] = app
            m[q][q] = aqq
            m[p][q] = 0f
            m[q][p] = 0f
            for (r in 0..2) {
                if (r != p && r != q) {
                    val arp = c * m[r][p] - s * m[r][q]
                    val arq = s * m[r][p] + c * m[r][q]
                    m[r][p] = arp; m[p][r] = arp
                    m[r][q] = arq; m[q][r] = arq
                }
            }
        }
        return listOf(m[0][0], m[1][1], m[2][2])
    }

    /** STUFE 2.2 (vereinfacht): RANSAC-Ebenenkonsens der dominanten Ebene. */
    fun ransacPlaneConsensus(
        points: List<Point3D>,
        distThreshold: Float = 0.05f,
    ): Pair<FloatArray, List<Point3D>> {
        if (points.size < 3) return FloatArray(3) to emptyList()
        var bestNormal = FloatArray(3)
        var bestInliers: List<Point3D> = emptyList()
        val iterations = minOf(80, maxOf(8, points.size / 2))
        for (iter in 0 until iterations) {
            val i1 = (iter * 31 + 7) % points.size
            val i2 = (iter * 57 + 13) % points.size
            val i3 = (iter * 91 + 23) % points.size
            if (i1 == i2 || i2 == i3 || i1 == i3) continue
            val p1 = points[i1]; val p2 = points[i2]; val p3 = points[i3]
            val ux = p2.x - p1.x; val uy = p2.y - p1.y; val uz = p2.z - p1.z
            val vx = p3.x - p1.x; val vy = p3.y - p1.y; val vz = p3.z - p1.z
            var nx = uy * vz - uz * vy
            var ny = uz * vx - ux * vz
            var nz = ux * vy - uy * vx
            val norm = sqrt(nx * nx + ny * ny + nz * nz)
            if (norm < 1e-9f) continue
            nx /= norm; ny /= norm; nz /= norm
            val d = -(nx * p1.x + ny * p1.y + nz * p1.z)
            val inliers = points.filter { abs(nx * it.x + ny * it.y + nz * it.z + d) < distThreshold }
            if (inliers.size > bestInliers.size) {
                bestInliers = inliers
                bestNormal = floatArrayOf(nx, ny, nz)
            }
        }
        return bestNormal to bestInliers
    }

    /** STUFE 3.1/3.3: Zylinder- + Plausibilitätsvalidierung (rein geometrisch). */
    fun validateDynamicCluster(points: List<Point3D>, centroid: Point3D): Boolean {
        if (points.size < CLUSTER_MIN_POINTS) return false
        val zMin = points.minOf { it.z }
        val zMax = points.maxOf { it.z }
        val height = zMax - zMin
        if (height < HEIGHT_MIN || height > HEIGHT_MAX) return false
        val width = points.maxOf { it.x } - points.minOf { it.x }
        if (width > MAX_WIDTH) return false
        // Zylinder: mind. 60 % der Punkte im Radius um das XY-Zentrum
        val inside = points.count {
            val dx = it.x - centroid.x
            val dy = it.y - centroid.y
            dx * dx + dy * dy <= CYLINDER_RADIUS * CYLINDER_RADIUS
        }
        if (inside < points.size * 0.6) return false
        val depth = points.maxOf { it.y } - points.minOf { it.y }
        val volume = width * depth * height
        if (volume < MIN_VOLUME) return false
        val dims = listOf(width, depth, height).sorted()
        val sphericity = dims.first() / (dims.last() + 1e-9f)
        return sphericity >= MIN_SPHERICITY
    }

    private fun centroid(points: List<Point3D>): Point3D =
        Point3D(
            points.map { it.x }.average().toFloat(),
            points.map { it.y }.average().toFloat(),
            points.map { it.z }.average().toFloat(),
        )

    private fun meanDistanceToOrigin(points: List<Point3D>): Float {
        if (points.isEmpty()) return 0f
        return points.map { sqrt(it.x * it.x + it.y * it.y + it.z * it.z) }.average().toFloat()
    }

    /**
     * Hauptklassifikationsmethode: liefert die Cluster-Berichte.
     * `persistable = false` ⇒ Live-Only (DYNAMIC), NIE speichern.
     */
    fun classify(points: List<Point3D>): Pair<List<ClusterReport>, Map<String, Int>> {
        if (points.isEmpty()) {
            return emptyList<ClusterReport>() to mapOf("total_points" to 0, "walls" to 0, "dynamic" to 0)
        }
        val down = voxelFilter(points)
        val filtered = heightFilter(down)
        val rawClusters = euclideanClustering(filtered).toMutableList()
        // Wände über 2,5 m als statische Struktur mitnehmen
        val high = points.filter { it.z > HEIGHT_MAX }
        if (high.size >= CLUSTER_MIN_POINTS) rawClusters.add(high)

        val reports = mutableListOf<ClusterReport>()
        var walls = 0
        var dynamic = 0
        for (cluster in rawClusters) {
            val planarity = calculatePlanarityScore(cluster)
            val dist = meanDistanceToOrigin(cluster)
            val threshold = if (dist < PLANARITY_DISTANCE_M) PLANARITY_THRESHOLD_NEAR else PLANARITY_THRESHOLD_FAR
            val c = centroid(cluster)
            val label: ObjectType
            val persistable: Boolean
            when {
                planarity > threshold -> { label = ObjectType.WALL; persistable = true; walls++ }
                validateDynamicCluster(cluster, c) -> { label = ObjectType.DYNAMIC; persistable = false; dynamic++ }
                else -> { label = ObjectType.WALL; persistable = true; walls++ } // zu flach/breit → Statik
            }
            reports.add(
                ClusterReport(
                    label = label,
                    centroid = c,
                    bboxMin = Point3D(
                        cluster.minOf { it.x }, cluster.minOf { it.y }, cluster.minOf { it.z },
                    ),
                    bboxMax = Point3D(
                        cluster.maxOf { it.x }, cluster.maxOf { it.y }, cluster.maxOf { it.z },
                    ),
                    planarity = planarity,
                    count = cluster.size,
                    persistable = persistable,
                    points = cluster,
                ),
            )
        }
        return reports to mapOf(
            "total_points" to points.size,
            "downsampled_points" to down.size,
            "height_filtered_points" to filtered.size,
            "clusters" to reports.size,
            "walls" to walls,
            "dynamic" to dynamic,
        )
    }

    /** Nur die persistierbaren Punkte (Statik) — für die Kartenspeicherung. */
    fun persistablePoints(points: List<Point3D>): List<Point3D> {
        val (reports, _) = classify(points)
        val kept = mutableListOf<Point3D>()
        for (report in reports) {
            if (report.persistable) kept.addAll(report.points)
        }
        kept.addAll(points.filter { it.z > HEIGHT_MAX }) // Decken-/Wandband
        return kept.distinctBy { listOf(it.x, it.y, it.z) }
    }
}

/** Punkt im Klassifikations-Koordinatensystem (Meter, Agent-Frame). */
data class Point3D(val x: Float, val y: Float, val z: Float)
